lib/web/sessions.go

```diff
@@ -497,148 +555,302 @@ func (s *sessionCache) ValidateTrustedCluster(validateRequest *auth.ValidateTrus
-		// close our extra context and return
-		if trace.IsAlreadyExists(err) {
-			defer c.Close()
-			return out, nil
+// Close releases all underlying resources for the user session.
+func (c *sessionResources) Close() error {
+	closers := c.transferClosers()
+	var errors []error
+	for _, closer := range closers {
+		c.log.Debugf("Closing %v.", closer)
+		if err := closer.Close(); err != nil {
+			errors = append(errors, err)
 		}
-		return nil, trace.Wrap(err)
 	}
-	return out, nil
+	return trace.NewAggregate(errors...)
 }
 
-func (s *sessionCache) SetSession(w http.ResponseWriter, user, sid string) error {
-	d, err := EncodeCookie(user, sid)
-	if err != nil {
-		return err
-	}
-	c := &http.Cookie{
-		Name:     "session",
-		Value:    d,
-		Path:     "/",
-		HttpOnly: true,
-		Secure:   true,
+// sessionResources persists resources initiated by a web session
+// but which might outlive the session.
+type sessionResources struct {
+	log logrus.FieldLogger
+
+	mu      sync.Mutex
+	closers []io.Closer
+}
+
+// addClosers adds the specified closers to this context
+func (c *sessionResources) addClosers(closers ...io.Closer) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.closers = append(c.closers, closers...)
+}
+
+// removeCloser removes the specified closer from this context
+func (c *sessionResources) removeCloser(closer io.Closer) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for i, cls := range c.closers {
+		if cls == closer {
+			c.closers = append(c.closers[:i], c.closers[i+1:]...)
+			return
+		}
 	}
-	http.SetCookie(w, c)
-	return nil
 }
 
-func (s *sessionCache) ClearSession(w http.ResponseWriter) {
-	http.SetCookie(w, &http.Cookie{
-		Name:     "session",
-		Value:    "",
-		Path:     "/",
-		HttpOnly: true,
-		Secure:   true,
+func (c *sessionResources) transferClosers() []io.Closer {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	closers := c.closers
+	c.closers = nil
+	return closers
+}
+
+func sessionKey(user, sessionID string) string {
+	return user + sessionID
+}
+
+// waitForWebSession will block until the requested web session shows up in the
+// cache or a timeout occurs.
+func (h *Handler) waitForWebSession(ctx context.Context, req proto.GetWebSessionRequest) error {
+	_, err := h.cfg.AccessPoint.GetWebSession(ctx, req)
+	if err == nil {
+		return nil
+	}
+	logger := h.log.WithField("req", req)
+	if !trace.IsNotFound(err) {
+		logger.WithError(err).Debug("Failed to query web session.")
+	}
+	// Establish a watch.
+	watcher, err := h.cfg.AccessPoint.NewWatcher(ctx, services.Watch{
+		Name: teleport.ComponentWebProxy,
+		Kinds: []services.WatchKind{
+			{
+				Kind:    services.KindWebSession,
+				SubKind: services.KindWebSession,
+			},
+		},
+		MetricComponent: teleport.ComponentWebProxy,
 	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer watcher.Close()
+	matchEvent := func(event services.Event) (services.Resource, error) {
+		if event.Type == backend.OpPut &&
+			event.Resource.GetKind() == services.KindWebSession &&
```

Comment:
Check subkind or that would be redundant?

Reply:
I'll add the subkind although the IDs should be unique between (even types of) sessions.